{"cells":[{"attachments":{},"cell_type":"markdown","metadata":{},"source":["# Projet phylogénétique\n","Marine Djaffardjy (marine.djaffardjy@u-psud.fr)- Théophile Sanchez - Sarah Cohen Boulakia\n","\n","------\n","\n","## Introduction\n","\n","Au cours de ce projet, vous étudierez trois espèces disparues de félins qui vivaient autrefois sur le continent Américain. Ces trois espèces, le _smilodon_ (tigre à dents de sabre), l'_homotherium_ (_scimitar toothed tigers_) et _M. trumani_ (guépard américain) se sont éteintes il y a environ 13 000 ans, à la fin de la dernière période glaciaire. Des séquences ADN partielles de la protéine cytochrome b de ces espèces ont pu être séquencées et vont vous permettre de retrouver les liens de parentés entre ces espèces et des espèces de félins contemporaines : le chat domestique, le lion, le léopard, le tigre, le puma, le guépard et les chats sauvages africains, chinois et européens. Sont aussi présentes dans le jeu de donnée des séquences issues d'espèces extérieures aux félins.\n","\n","Afin de reconstruire l'arbre phylogénétique de ces espèces, vous utiliserez une méthode basée sur le calcul des distances évolutives entre les séquences ADN des protéines. Sachez qu'une démarche similaire peut-être appliquée aux séquences d'acides aminés.\n","\n","Les différentes étapes qui vous permetterons de construire l'arbre sont détaillées dans ce notebook. Vous devrez implémenter les algorithmes en Python et répondre aux questions dans les cellules prévues.\n","\n","Quelques conseils :\n","- Utiliser au maximum les fonctions présentes dans les packages de python (sauf si il vous est explicitement demandé de les réimplémenter). Si un problème vous paraît courant, il existe surement déjà une fonction pour le résoudre. Pour ce projet vous serez limités aux packages de base, à Numpy et ETE (seulement pour l'affichage des arbres). Vous n'avez pas besoin d'utiliser d'autre packages (pas de pandas).\n","- Si une partie de votre code ne vous semble pas très explicite, ajoutez des commentaires pour l'expliquer. Une personne qui lit votre code doit pouvoir comprendre son fonctionnement facilement.\n","- Faites des fonctions d'affichage pour vos matrices !!\n","- N'hésitez pas à chercher dans la documentation et sur internet. Cependant, faites attention au plagiat !\n","\n","Le projet est à rendre **en binôme** par mail. Vous regrouperez votre notebook et les fichiers nécessaires à son fonctionnement dans une archive portant vos noms et prénoms (il faut que je puisse faire tourner les notebooks)."]},{"cell_type":"markdown","metadata":{},"source":["------\n","## Importation des séquences\n","\n","Le format [FASTA](https://fr.wikipedia.org/wiki/FASTA_(format_de_fichier) permet de stocker plusieurs séquences (ADN, ARN ou peptidiques) dans un fichier. Les séquences que vous allez étudier ont été regroupées dans le fichier `cat_dna.fasta`.\n","\n","**Exercice 1 :** Écrivez une fonction permettant d'importer un fichier au format fasta et de le stocker dans un dictionnaire `seq_cat`. Les clés seront les noms des séquences et les valeurs du dictionnaire seront les séquences d'adn.\n","Exportez le dictionnaire obtenu dans un fichier `output/dict_seq.json`."]},{"cell_type":"code","execution_count":11,"metadata":{"trusted":true},"outputs":[],"source":["\n","def load_dict(filename):\n","    \"\"\"\n","    Importe un fichier au format FASTA et le stocke dans un dictionnaire.\n","    Les clés sont les noms des séquences et les valeurs sont les séquences d'ADN.\n","\n","    Args:\n","        filename (str): Le nom du fichier à importer.\n","\n","    Returns:\n","        dict: Le dictionnaire des séquences.\n","    \"\"\"\n","    seq_cat = {}\n","    file = open(filename)\n","    value=\"\"\n","    for i, line in enumerate(file):\n","        if line[0]==\">\":\n","            label = line[1:]\n","        elif len(line)>1:\n","            value+=line.strip()\n","        else:\n","            seq_cat[label.strip()]=value\n","            value=\"\"\n","\n","    # Exportation du dictionnaire dans un fichier JSON\n","    with open(\"output/dict_seq.json\", \"w\") as f:\n","        f.write(str(seq_cat))\n","\n","    return seq_cat\n","\n","# Exemple d'utilisation\n","#load_dict(\"input/cat_dna.fasta\")"]},{"cell_type":"markdown","metadata":{},"source":["------\n","## Alignement des séquences\n","\n","La méthode que vous utiliserez pour calculer l'arbre phylogénétique nécessite de calculer la distance évolutive entre les séquences. Avant de pouvoir les calculer, il faut d'abord aligner les séquences en considérant trois types de mutations :\n","- les substitutions (un nucléotide est remplacé par un autre)\n","- les insertions\n","- les délétions\n","Par exemple, les séquences \"ACTCCTGA\" et \"ATCTCGTGA\" ont plusieurs alignements possibles : \n","\n","$A_1$ :\n","```\n","-ACTCCTGA\n","ATCTCGTGA\n","```\n","\n","$A_2$ :\n","```\n","A-CTCCTGA\n","ATCTCGTGA\n","```\n","\n","$A_3$ :\n","```\n","AC-TCCTGA\n","ATCTCGTGA\n","```\n",".\n","\n",".\n","\n",".\n","\n","Le \"-\" désigne un *gap*, c'est à dire un \"trou\" dans l'alignement qui a été causé par une insertion ou une déletion. On regroupe ces deux types de mutations sous le terme indel.\n","\n","Ces alignements correspondent à une multitude d'histoires phylogénétiques différentes. Pour sélectionner le meilleur alignement il faut donc introduire l'hypothèse du maximum de parcimonie qui privilégie l'histoire phylogénétique qui implique le moins d'hypothèses et donc, le moins de changements évolutifs. Par exemple, parmis les trois alignements ci-dessus on preferera l'alignement 2 car il correspond au scénario avec le moins de mutations:\n","- l'alignement 1 implique au minimum 1 indel et 2 substitutions\n","- l'alignement 2 implique au minimum 1 indel et 1 substitutions\n","- l'alignement 3 implique au minimum 1 indel et 2 substitutions\n","\n","On peut maintenant définir un score d'identité que l'on va augmenter de 1 lorsque qu'il n'y pas eu de mutation et ainsi obtenir la matrice suivante :\n","\n","|   &nbsp;   | A | C | G | T | - |\n","|   -   | - | - | - | - | - |\n","| **A** | 1 | 0 | 0 | 0 | 0 |\n","| **C** | 0 | 1 | 0 | 0 | 0 |\n","| **G** | 0 | 0 | 1 | 0 | 0 |\n","| **T** | 0 | 0 | 0 | 1 | 0 |\n","| **-** | 0 | 0 | 0 | 0 | 0 |\n","\n","Cette matrice correspond au modèle d'évolution de l'ADN défini par Jukes et Cantor qui fait l'hypothèse d'un taux de mutation équivalent pour chacun des nucléotides. Cependant, en réalité ces taux ne sont pas les mêmes partout, on sait par exemple que le taux de transition (substitution A$\\leftrightarrow$G ou C$\\leftrightarrow$T) est différent du taux de transversions (substitution A$\\leftrightarrow$T, C$\\leftrightarrow$G, C$\\leftrightarrow$A ou G$\\leftrightarrow$T) et que d'autres facteurs devrait être pris en compte comme la fréquence du nucléotide dans l'ADN. [C'est pour cette raison qu'il existe beaucoup de modèles différents d'écrivant l'évolution de l'ADN.](https://en.wikipedia.org/wiki/Models_of_DNA_evolution) Dans la suite de ce projet nous utiliserons la matrice de similarité $S$ suivante : \n","\n","|   &nbsp;   | A  | C  | G  | T  | -  |\n","|   -   | -  | -  | -  | -  | -  |\n","| **A** | 10 | -1 | -3 | -4 | -5 |\n","| **C** | -1 | 7  | -5 | -3 | -5 |\n","| **G** | -3 | -5 | 9  | 0  | -5 |\n","| **T** | -4 | -3 | 0  | 8  | -5 |\n","| **-** | -5 | -5 | -5 | -5 | 0  |\n","\n","**Exercice 2 :** Écriver la fonction permettant de calculer le score entre deux alignements avec la matrice de similarité précédente puis afficher le score des trois alignements $A_1$, $A_2$ et $A_3$. La classe permettant d'importer une matrice et de calculer le score entre deux lettres vous est déjà fournie, la matrice de similarité est stockée dans le fichier `dna_matrix` :\n"]},{"cell_type":"code","execution_count":2,"metadata":{"trusted":true},"outputs":[{"name":"stdout","output_type":"stream","text":["Score between G and C: -5.0\n","Score between A and a gap: -5.0\n"]}],"source":["import numpy as np\n","\n","class SimilarityMatrix:\n","    def __init__(self, filename):\n","        with open(filename) as f:\n","            self.letters = f.readline().split()\n","            self.values = np.loadtxt(filename, skiprows=1, usecols=range(1, len(self.letters) + 1))\n","        \n","    def score(self, letter1, letter2): # return the similarity score between letter1 and letter2\n","        return self.values[self.letters.index(letter1)][self.letters.index(letter2)]\n","    \n","# Example\n","similarity_matrix = SimilarityMatrix('input/dna_matrix')\n","print('Score between G and C:', similarity_matrix.score('G', 'C'))\n","print('Score between A and a gap:', similarity_matrix.score('-', 'A'))"]},{"cell_type":"code","execution_count":3,"metadata":{"trusted":true},"outputs":[{"name":"stdout","output_type":"stream","text":["35.0\n","49.0\n","39.0\n"]}],"source":["similarity_matrix = SimilarityMatrix('input/dna_matrix')\n","\n","def score_compare(a,b):\n","    score=0\n","    for i in range(len(a)):\n","            score+=similarity_matrix.score(a[i],b[i])\n","    return score\n","\n","\n","print(score_compare(\"-ACTCCTGA\",\"ATCTCGTGA\")) #A1\n","print(score_compare(\"A-CTCCTGA\",\"ATCTCGTGA\")) #A2\n","print(score_compare(\"AC-TCCTGA\",\"ATCTCGTGA\")) #A3"]},{"cell_type":"markdown","metadata":{},"source":["------\n","### Algorithme de Needleman-Wunsch\n","\n","Maintenant que vous avez vu ce qu'est une matrice de similarité et comment calculer le score de similarité d'un alignement, vous allez devoir implémenter un algorithme permettant de trouver le meilleur alignement global entre deux séquences. Avec deux séquences à aligner de taille $n$ et $m$, la première étape consiste à initialiser deux matrices de taille $(n+1) \\times (m+1)$. La première est la matrice de score $M$ et la seconde sera la matrice de *traceback* $T$. \n","\n","Par exemple, avec la matrice $S$ et les séquences $A =$ \"ACTCCTGA\" et $B =$ \"ATCTCGTGA\", on initialise $M$ comme si l'on ajoutait des *gaps* partout :\n","\n","|   &nbsp;   | - | A | T | C | T | C | G | T | G | A |\n","|   -   | - | - | - | - | - | - | - | - | - | - |\n","| **-** | 0 |-5 |-10|-15|-20|-25|-30|-35|-40|-45|\n","| **A** |-5 | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; |\n","| **C** |-10| &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; |\n","| **T** |-15| &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; |\n","| **C** |-20| &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; |\n","| **C** |-25| &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; |\n","| **T** |-30| &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; |\n","| **G** |-45| &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; |\n","| **A** |-40| &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; ||\n","\n","Puis on initialise $T$ :\n","\n","|   &nbsp;   | - | A | T | C | T | C | G | T | G | A |\n","|   -   | - | - | - | - | - | - | - | - | - | - |\n","| **-** | o | l | l | l | l | l | l | l | l | l |\n","| **A** | u | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; |\n","| **C** | u | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; |\n","| **T** | u | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; |\n","| **C** | u | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; |\n","| **C** | u | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; |\n","| **T** | u | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; |\n","| **G** | u | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; |\n","| **A** | u | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; ||\n","\n","\n","Il faut ensuite remplir la matrice $M$ en suivant la formule $M_{ij} = \\max(M_{i-1j-1} + s(A_i, B_j), M_{ij-1} + s(A_i, gap), M_{i-1j} + s(B_j,gap) )$ avec $i \\in {2, \\dots, n+1}$, $j \\in {2, \\dots, m+1}$ et la fonction $s$ qui calcule le score de similarité entre deux nucléotides. Pour chaque case de $T$ on remplie par :\n","- 'd' (*diagonal*) si $M_{ij}$ a été calculé en utilisant la diagonale $M_{i-1j-1}$,\n","- 'l' (*left*) si $M_{ij}$ a été calculé en utilisant la case de gauche $M_{ij-1}$,\n","- 'u' (*up*) si $M_{ij}$ a été calculé en utilisant la case du haut $M_{i-1j}$.\n","\n","On obtient alors les matrices suivantes $M$ et $T$ : \n","\n","|   &nbsp;   | - | A | T | C | T | C | G | T | G | A |\n","|   -   | - | - | - | - | - | - | - | - | - | - |\n","| **-** |  0| -5|-10|-15|-20|-25|-30|-35|-40|-45|\n","| **A** | -5| 10|  5|  0| -5|-10|-15|-20|-25|-30|\n","| **C** |-10|  5|  7| 12|  7|  2| -3| -8|-13|-18|\n","| **T** |-15|  0| 13|  8| 20| 15| 10|  5|  0| -5|\n","| **C** |-20| -5|  8| 20| 15| 27| 22| 17| 12|  7|\n","| **C** |-25|-10|  3| 15| 17| 22| 22| 19| 14| 11|\n","| **T** |-30|-15| -2| 10| 23| 18| 22| 30| 25| 20|\n","| **G** |-35|-20| -7|  5| 18| 18| 27| 25| 39| 34|\n","| **A** |-40|-25|-12|  0| 13| 17| 22| 23| 34| 49|\n","\n","|   &nbsp;   | - | A | T | C | T | C | G | T | G | A |\n","|   -   | - | - | - | - | - | - | - | - | - | - |\n","| **-** | o | l | l | l | l | l | l | l | l | l |\n","| **A** | u | d | l | l | l | l | l | l | l | d |\n","| **C** | u | u | d | d | l | d | l | l | l | l |\n","| **T** | u | u | d | l | d | l | l | d | l | l |\n","| **C** | u | u | u | d | l | d | l | l | l | l |\n","| **C** | u | u | u | d | d | d | d | d | l | d |\n","| **T** | u | u | d | u | d | l | d | d | l | l |\n","| **G** | u | u | u | u | u | d | d | u | d | l |\n","| **A** | u | d | u | u | u | d | u | d | u | d |\n","\n","Il suffit maintenant de regarder le dernier élément $M_{nm} = 49$ pour avoir le score de l'alignement. Pour avoir l'alignement lui-même, il faut partir de $T_{nm}$ et remonter la \"trace\" jusqu'à arriver au 'o'. Un 'd' correspond à un *match* entre les deux séquences, 'l' à un *gap* dans la séquence $A$ et 'u' à un *gap* dans la séquence $B$. En revenant à l'exemple précédent on obtient la trace suivante :\n","\n","|   &nbsp;   | - | A | T | C | T | C | G | T | G | A |\n","|   -   | - | - | - | - | - | - | - | - | - | - |\n","| **-** | o | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; |\n","| **A** | &nbsp; | d | l | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; |\n","| **C** | &nbsp; | &nbsp; | &nbsp; | d | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; |\n","| **T** | &nbsp; | &nbsp; | &nbsp; | &nbsp; | d | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; |\n","| **C** | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | d | &nbsp; | &nbsp; | &nbsp; | &nbsp; |\n","| **C** | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | d | &nbsp; | &nbsp; | &nbsp; |\n","| **T** | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | d | &nbsp; | &nbsp; |\n","| **G** | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | d | &nbsp; |\n","| **A** | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | d |\n","\n","Elle correspond à l'alignement :\n","```\n","A-CTCCTGA\n","ATCTCGTGA\n","```\n","\n","**Exercice 3 :** Implémenter l'algorithme de Needlman et Wunsch. Il prendra en paramètre deux séquences et une matrice de similarité et retournera leur alignement. Tester le avec les séquences \"ACTCCTGA\" et \"ATCTCGTGA\"."]},{"cell_type":"code","execution_count":4,"metadata":{"trusted":true},"outputs":[],"source":["def algo(a,b,mat):\n","    \n","    #Initialisation des matrices\n","    n, m = len(a), len(b)\n","    M = np.zeros((n+1,m+1))\n","    T = np.zeros((n+1,m+1))\n","\n","    for i in range (1,n+1):\n","        M[i][0] = -5*i\n","        T[i][0]= 1\n","        \n","    for i in range (1,m+1):\n","        M[0][i] = -5*i\n","        T[0][i]= 2\n","        \n","    #Algo\n","    \n","    for i in range (n):\n","        for j in range (m):\n","            M[i+1][j+1] = max ( M[i][j] + mat.score(a[i],b[j]),\n","                            M[i][j+1] + mat.score(a[i],\"-\"),\n","                            M[i+1][j] + mat.score(b[j],\"-\"))\n","\n","            if max( M[i][j] + mat.score(a[i],b[j]), M[i][j+1] + mat.score(a[i],\"-\"),M[i+1][j] + mat.score(b[j],\"-\"))==M[i][j+1] + mat.score(a[i],\"-\"):\n","                T[i+1][j+1]=1\n","            if max( M[i][j] + mat.score(a[i],b[j]), M[i][j+1] + mat.score(a[i],\"-\"),M[i+1][j] + mat.score(b[j],\"-\"))==M[i+1][j] + mat.score(b[j],\"-\"):\n","                T[i+1][j+1]=2\n","            if max(M[i][j] + mat.score(a[i],b[j]), M[i][j+1] + mat.score(a[i],\"-\"),M[i+1][j] + mat.score(b[j],\"-\"))==M[i][j] + mat.score(a[i],b[j]):\n","                T[i+1][j+1]=3\n","    \n","    S1 = \"\"\n","    S2 = \"\"\n","    while(T[n][m] != 0):\n","        if T[n][m] == 3 :\n","            S1 += a[n-1]\n","            S2 += b[m-1]\n","            n = n - 1\n","            m = m - 1\n","        if T[n][m] == 2 :\n","            S1 += \"-\"\n","            S2 += b[m-1]\n","            m = m - 1\n","        if T[n][m] == 1 :\n","            S1 += a[n-1]\n","            S2 += \"-\"\n","            n = n - 1 \n","    S1 = \"\".join(reversed(S1))\n","    S2 = \"\".join(reversed(S2))\n","    L = [S1, S2]          \n","            \n","    return S1,S2\n","\n","\n","A,B=algo(\"ACTCCTGA\",\"ATCTCGTGA\",similarity_matrix)\n","\n","# print(A)\n","# for i in B:\n","#     a=\"\"\n","#     for j in i:\n","#         if int(j)==0:\n","#             a+=\"o  \"\n","#         elif int(j)==1:\n","#             a+=\"u  \"\n","#         elif int(j)==2:\n","#             a+=\"l  \"\n","#         elif int(j)==3:\n","#             a+=\"d  \"\n","#     print(a)\n"]},{"cell_type":"markdown","metadata":{},"source":["----\n","## Matrice de distance\n","\n","Dans le cas de séquences très proches, on estime que la distance évolutive réelle entre les séquences est proche de la p-distance qui est simplement le nombre de substitution dans l'alignement sur le nombre total de nucléotide. Pour simplifier, on ignore les positions alignées à des gaps. On applique ensuite la correction de Jukes-Cantor. Sa formule est $-(\\frac{3}{4})\\ln(1-(\\frac{4}{3})\\times \\textit{p-distance})$.\n","\n","**Exercice 4 :** Implémenter la fonction retournant la matrice de distance à partir d'un dictionnaire de séquences. "]},{"cell_type":"code","execution_count":5,"metadata":{"trusted":true},"outputs":[{"name":"stdout","output_type":"stream","text":["          |  Saber  |  Homot  |  Ameri  |  Spott  |  Strip  |  Chine  |  Afric  |  Felis  |  Wild   |  Cheet  |  Puma   |  Tiger  |  Lion   |  Leopa  |  Dog D  |  Wolf   |  Black  |  Brown  |  Cave  \n","\n"," |  Saber \n","\n"," |  Homot  |  0.140 \n","\n"," |  Ameri  |  0.160  |  0.170 \n","\n"," |  Spott  |  0.131  |  0.145  |  0.167 \n","\n"," |  Strip  |  0.175  |  0.176  |  0.162  |  0.127 \n","\n"," |  Chine  |  0.135  |  0.166  |  0.102  |  0.147  |  0.161 \n","\n"," |  Afric  |  0.152  |  0.181  |  0.110  |  0.170  |  0.153  |  0.098 \n","\n"," |  Felis  |  0.157  |  0.176  |  0.130  |  0.166  |  0.167  |  0.110  |  0.018 \n","\n"," |  Wild   |  0.157  |  0.185  |  0.114  |  0.162  |  0.162  |  0.102  |  0.011  |  0.014 \n","\n"," |  Cheet  |  0.157  |  0.176  |  0.078  |  0.175  |  0.132  |  0.115  |  0.106  |  0.118  |  0.110 \n","\n"," |  Puma   |  0.187  |  0.206  |  0.055  |  0.198  |  0.202  |  0.123  |  0.139  |  0.160  |  0.143  |  0.090 \n","\n"," |  Tiger  |  0.139  |  0.192  |  0.126  |  0.145  |  0.185  |  0.123  |  0.114  |  0.127  |  0.118  |  0.115  |  0.118 \n","\n"," |  Lion   |  0.144  |  0.198  |  0.152  |  0.176  |  0.181  |  0.135  |  0.140  |  0.153  |  0.144  |  0.124  |  0.136  |  0.106 \n","\n"," |  Leopa  |  0.143  |  0.201  |  0.152  |  0.176  |  0.189  |  0.136  |  0.124  |  0.128  |  0.128  |  0.150  |  0.153  |  0.130  |  0.078 \n","\n"," |  Dog D  |  0.204  |  0.232  |  0.231  |  0.210  |  0.202  |  0.205  |  0.227  |  0.208  |  0.222  |  0.199  |  0.240  |  0.228  |  0.222  |  0.203 \n","\n"," |  Wolf   |  0.204  |  0.232  |  0.231  |  0.210  |  0.202  |  0.205  |  0.227  |  0.208  |  0.222  |  0.199  |  0.240  |  0.228  |  0.222  |  0.203  |  -0.000 \n","\n"," |  Black  |  0.204  |  0.229  |  0.191  |  0.218  |  0.201  |  0.195  |  0.204  |  0.217  |  0.218  |  0.191  |  0.185  |  0.200  |  0.169  |  0.191  |  0.215  |  0.215 \n","\n"," |  Brown  |  0.213  |  0.229  |  0.228  |  0.220  |  0.202  |  0.205  |  0.242  |  0.238  |  0.247  |  0.219  |  0.195  |  0.195  |  0.186  |  0.209  |  0.201  |  0.201  |  0.127 \n","\n"," |  Cave   |  0.222  |  0.191  |  0.190  |  0.187  |  0.192  |  0.204  |  0.209  |  0.204  |  0.213  |  0.191  |  0.185  |  0.159  |  0.187  |  0.209  |  0.180  |  0.180  |  0.102  |  0.070 \n","\n"]},{"data":{"text/plain":["array([[ 0.        ,  0.13993947,  0.16034827,  0.13144079,  0.17493124,\n","         0.13463146,  0.15227189,  0.15661344,  0.15661344,  0.1572313 ,\n","         0.18697828,  0.13885682,  0.14366347,  0.14310613,  0.20395029,\n","         0.20395029,  0.20395029,  0.21338437,  0.22201531],\n","       [ 0.13993947,  0.        ,  0.17046694,  0.14536195,  0.17633981,\n","         0.16602905,  0.18087154,  0.17563268,  0.18468257,  0.17633981,\n","         0.20610665,  0.19228138,  0.1984655 ,  0.20146985,  0.23164626,\n","         0.23164626,  0.22871628,  0.22871628,  0.19079352],\n","       [ 0.16034827,  0.17046694,  0.        ,  0.1666907 ,  0.16161728,\n","         0.10165912,  0.10976731,  0.13042977,  0.11428909,  0.07813751,\n","         0.05496965,  0.12625149,  0.15227189,  0.15227189,  0.23068445,\n","         0.23068445,  0.19079352,  0.22776181,  0.190018  ],\n","       [ 0.13144079,  0.14536195,  0.1666907 ,  0.        ,  0.12673751,\n","         0.14737964,  0.17046694,  0.16602905,  0.16161728,  0.17493124,\n","         0.19766025,  0.14536195,  0.17563268,  0.17563268,  0.20951524,\n","         0.20951524,  0.21814629,  0.21996998,  0.18694578],\n","       [ 0.17493124,  0.17633981,  0.16161728,  0.12673751,  0.        ,\n","         0.16098026,  0.1534745 ,  0.1666907 ,  0.16225937,  0.13246765,\n","         0.20228989,  0.18468257,  0.18087154,  0.18924879,  0.2017536 ,\n","         0.2017536 ,  0.20092134,  0.2017536 ,  0.19157542],\n","       [ 0.13463146,  0.16602905,  0.10165912,  0.14737964,  0.16098026,\n","         0.        ,  0.09800623,  0.11018511,  0.10204392,  0.11472701,\n","         0.12256504,  0.12303734,  0.13515274,  0.13567809,  0.20479081,\n","         0.20479081,  0.1946334 ,  0.20479081,  0.20395029],\n","       [ 0.15227189,  0.18087154,  0.10976731,  0.17046694,  0.1534745 ,\n","         0.09800623,  0.        ,  0.01781556,  0.01063848,  0.10610347,\n","         0.13885682,  0.11428909,  0.13993947,  0.12351331,  0.22681533,\n","         0.22681533,  0.20395029,  0.24242026,  0.2086525 ],\n","       [ 0.15661344,  0.17563268,  0.13042977,  0.16602905,  0.1666907 ,\n","         0.11018511,  0.01781556,  0.        ,  0.01421844,  0.11841565,\n","         0.16034827,  0.12673751,  0.15287082,  0.1277209 ,  0.20779688,\n","         0.20779688,  0.21724582,  0.23750221,  0.20395029],\n","       [ 0.15661344,  0.18468257,  0.11428909,  0.16161728,  0.16225937,\n","         0.10204392,  0.01063848,  0.01421844,  0.        ,  0.11018511,\n","         0.14310613,  0.11841565,  0.14422519,  0.1277209 ,  0.22201531,\n","         0.22201531,  0.21814629,  0.24737077,  0.21338437],\n","       [ 0.1572313 ,  0.17633981,  0.07813751,  0.17493124,  0.13246765,\n","         0.11472701,  0.10610347,  0.11841565,  0.11018511,  0.        ,\n","         0.08999547,  0.1151683 ,  0.12351331,  0.14970979,  0.19927737,\n","         0.19927737,  0.19079352,  0.21905432,  0.19079352],\n","       [ 0.18697828,  0.20610665,  0.05496965,  0.19766025,  0.20228989,\n","         0.12256504,  0.13885682,  0.16034827,  0.14310613,  0.08999547,\n","         0.        ,  0.11841565,  0.13567809,  0.15287082,  0.24039332,\n","         0.24039332,  0.18543083,  0.1946334 ,  0.18468257],\n","       [ 0.13885682,  0.19228138,  0.12625149,  0.14536195,  0.18468257,\n","         0.12303734,  0.11428909,  0.12673751,  0.11841565,  0.1151683 ,\n","         0.11841565,  0.        ,  0.10570226,  0.13042977,  0.22776181,\n","         0.22776181,  0.20009596,  0.19543032,  0.1591146 ],\n","       [ 0.14366347,  0.1984655 ,  0.15227189,  0.17563268,  0.18087154,\n","         0.13515274,  0.13993947,  0.15287082,  0.14422519,  0.12351331,\n","         0.13567809,  0.10570226,  0.        ,  0.0778476 ,  0.22201531,\n","         0.22201531,  0.1687078 ,  0.18618521,  0.18694578],\n","       [ 0.14310613,  0.20146985,  0.15227189,  0.17563268,  0.18924879,\n","         0.13567809,  0.12351331,  0.1277209 ,  0.1277209 ,  0.14970979,\n","         0.15287082,  0.13042977,  0.0778476 ,  0.        ,  0.20311668,\n","         0.20311668,  0.19079352,  0.2086525 ,  0.2086525 ],\n","       [ 0.20395029,  0.23164626,  0.23068445,  0.20951524,  0.2017536 ,\n","         0.20479081,  0.22681533,  0.20779688,  0.22201531,  0.19927737,\n","         0.24039332,  0.22776181,  0.22201531,  0.20311668,  0.        ,\n","        -0.        ,  0.21546715,  0.20146985,  0.18014398],\n","       [ 0.20395029,  0.23164626,  0.23068445,  0.20951524,  0.2017536 ,\n","         0.20479081,  0.22681533,  0.20779688,  0.22201531,  0.19927737,\n","         0.24039332,  0.22776181,  0.22201531,  0.20311668, -0.        ,\n","         0.        ,  0.21546715,  0.20146985,  0.18014398],\n","       [ 0.20395029,  0.22871628,  0.19079352,  0.21814629,  0.20092134,\n","         0.1946334 ,  0.20395029,  0.21724582,  0.21814629,  0.19079352,\n","         0.18543083,  0.20009596,  0.1687078 ,  0.19079352,  0.21546715,\n","         0.21546715,  0.        ,  0.12673751,  0.10204392],\n","       [ 0.21338437,  0.22871628,  0.22776181,  0.21996998,  0.2017536 ,\n","         0.20479081,  0.24242026,  0.23750221,  0.24737077,  0.21905432,\n","         0.1946334 ,  0.19543032,  0.18618521,  0.2086525 ,  0.20146985,\n","         0.20146985,  0.12673751,  0.        ,  0.0700755 ],\n","       [ 0.22201531,  0.19079352,  0.190018  ,  0.18694578,  0.19157542,\n","         0.20395029,  0.2086525 ,  0.20395029,  0.21338437,  0.19079352,\n","         0.18468257,  0.1591146 ,  0.18694578,  0.2086525 ,  0.18014398,\n","         0.18014398,  0.10204392,  0.0700755 ,  0.        ]])"]},"execution_count":5,"metadata":{},"output_type":"execute_result"}],"source":["def MatDist(Dic):\n","    K = list(Dic.keys())\n","    V = list(Dic.values())\n","    M = np.zeros((len(K),len(K)))\n","    print(\"        \", end = \" \")\n","    for i in range(len(K)):\n","        print(\" | \", f\"{K[i]:5.5}\", end  = \" \")\n","    print(\"\\n\")\n","    for i in range(len(K)):\n","        print(\" | \", f\"{K[i]:5.5}\", end = \" \")\n","        for j in range(0, i):\n","            A,B=algo(V[i],V[j],similarity_matrix)\n","            M[i][j] = p_distance(A,B)\n","            M[j][i] = p_distance(A,B)\n","            print(\" | \", \"{0:.3f}\".format(M[i][j]), end = \" \")\n","        print(\"\\n\")\n","    return M\n","\n","def p_distance(Seq1, Seq2):\n","    count = 0\n","    countGap = 0\n","    a = len(Seq1)\n","    b = len(Seq2)\n","    for i in range(a) :\n","        if Seq1[i] != Seq2[i] and Seq1[i] != '-' and Seq2[i] != '-':\n","            count += 1\n","        if Seq1[i] == '-':\n","            countGap += 1\n","        if Seq2[i] == '-':\n","            countGap += 1\n","    count = count/(len(Seq1)-countGap)\n","    count = -(3/4)*np.log(1-(4/3)*count)\n","    return count\n","    \n","Dic = load_dict(\"input/cat_dna.fasta\")\n","MatDist(Dic)"]},{"cell_type":"markdown","metadata":{},"source":["**Exercice 4 bis :** À quoi sert la correction de Jukes-Cantor ? Quelle hypothèse fait-elle ? "]},{"cell_type":"markdown","metadata":{},"source":["Réponse :"]},{"cell_type":"markdown","metadata":{},"source":["------\n","## Construction d'un arbre avec UPGMA\n","\n","Grâce aux mesures de distances entre les séquences, on peut maintenant de construire l'arbre phylogénétique des globines. Vous allez devoir pour cela implémenter l'algorithme UPGMA (*unweighted pair group method with arithmetic mean*) qui, malgré son nom compliqué, est l'une des méthodes les plus simples pour la construction d'arbre.\n","\n","### Le format Newick\n","\n","Le format Newick est l'un des formats utilisé en phylogénie pour représenter un arbre sous la forme d'une chaine de caractère. Le principe est simple, les groupes ayant la même racine sont écrit entre parenthèses et séparés par des virgules. Un groupe peut être soit une feuille de l'arbre (dans notre cas une séquence), soit un autre groupe. La longueur de la branche de chaque groupe est écrite après un double point et l'arbre est terminé par un point virgule. Pour afficher l'arbre on peut utiliser les fonction du package ETE : "]},{"cell_type":"code","execution_count":6,"metadata":{"trusted":true},"outputs":[{"ename":"TypeError","evalue":"arguments did not match any overloaded call:\n  QImage(): too many arguments\n  QImage(size: QSize, format: QImage.Format): argument 1 has unexpected type 'float'\n  QImage(width: int, height: int, format: QImage.Format): argument 1 has unexpected type 'float'\n  QImage(data: bytes, width: int, height: int, format: QImage.Format): argument 1 has unexpected type 'float'\n  QImage(data: PyQt5.sip.voidptr, width: int, height: int, format: QImage.Format): argument 1 has unexpected type 'float'\n  QImage(data: bytes, width: int, height: int, bytesPerLine: int, format: QImage.Format): argument 1 has unexpected type 'float'\n  QImage(data: PyQt5.sip.voidptr, width: int, height: int, bytesPerLine: int, format: QImage.Format): argument 1 has unexpected type 'float'\n  QImage(xpm: List[str]): argument 1 has unexpected type 'float'\n  QImage(fileName: str, format: typing.Optional[str] = None): argument 1 has unexpected type 'float'\n  QImage(a0: QImage): argument 1 has unexpected type 'float'\n  QImage(variant: Any): too many arguments","output_type":"error","traceback":["\u001b[1;31m---------------------------------------------------------------------------\u001b[0m","\u001b[1;31mTypeError\u001b[0m                                 Traceback (most recent call last)","Cell \u001b[1;32mIn[6], line 7\u001b[0m\n\u001b[0;32m      5\u001b[0m ts \u001b[39m=\u001b[39m TreeStyle()\n\u001b[0;32m      6\u001b[0m ts\u001b[39m.\u001b[39mshow_branch_length \u001b[39m=\u001b[39m \u001b[39mTrue\u001b[39;00m\n\u001b[1;32m----> 7\u001b[0m t\u001b[39m.\u001b[39;49mrender(\u001b[39m'\u001b[39;49m\u001b[39m%%\u001b[39;49;00m\u001b[39minline\u001b[39;49m\u001b[39m'\u001b[39;49m, w\u001b[39m=\u001b[39;49m\u001b[39m183\u001b[39;49m, units\u001b[39m=\u001b[39;49m\u001b[39m'\u001b[39;49m\u001b[39mmm\u001b[39;49m\u001b[39m'\u001b[39;49m, tree_style\u001b[39m=\u001b[39;49mts)\n","File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python310\\site-packages\\ete3\\coretype\\tree.py:1392\u001b[0m, in \u001b[0;36mTreeNode.render\u001b[1;34m(self, file_name, layout, w, h, tree_style, units, dpi)\u001b[0m\n\u001b[0;32m   1388\u001b[0m     \u001b[39mreturn\u001b[39;00m drawer\u001b[39m.\u001b[39mget_img(\u001b[39mself\u001b[39m, w\u001b[39m=\u001b[39mw, h\u001b[39m=\u001b[39mh,\n\u001b[0;32m   1389\u001b[0m                           layout\u001b[39m=\u001b[39mlayout, tree_style\u001b[39m=\u001b[39mtree_style,\n\u001b[0;32m   1390\u001b[0m                           units\u001b[39m=\u001b[39munits, dpi\u001b[39m=\u001b[39mdpi, return_format\u001b[39m=\u001b[39mfile_name)\n\u001b[0;32m   1391\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[1;32m-> 1392\u001b[0m     \u001b[39mreturn\u001b[39;00m drawer\u001b[39m.\u001b[39;49mrender_tree(\u001b[39mself\u001b[39;49m, file_name, w\u001b[39m=\u001b[39;49mw, h\u001b[39m=\u001b[39;49mh,\n\u001b[0;32m   1393\u001b[0m                             layout\u001b[39m=\u001b[39;49mlayout, tree_style\u001b[39m=\u001b[39;49mtree_style,\n\u001b[0;32m   1394\u001b[0m                               units\u001b[39m=\u001b[39;49munits, dpi\u001b[39m=\u001b[39;49mdpi)\n","File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python310\\site-packages\\ete3\\treeview\\drawer.py:111\u001b[0m, in \u001b[0;36mrender_tree\u001b[1;34m(t, imgName, w, h, layout, tree_style, header, units, dpi)\u001b[0m\n\u001b[0;32m    109\u001b[0m scene\u001b[39m.\u001b[39maddItem(scene\u001b[39m.\u001b[39mmaster_item)\n\u001b[0;32m    110\u001b[0m \u001b[39mif\u001b[39;00m imgName\u001b[39m.\u001b[39mstartswith(\u001b[39m\"\u001b[39m\u001b[39m%%\u001b[39;00m\u001b[39minline\u001b[39m\u001b[39m\"\u001b[39m):\n\u001b[1;32m--> 111\u001b[0m     imgmap \u001b[39m=\u001b[39m save(scene, imgName, w\u001b[39m=\u001b[39;49mw, h\u001b[39m=\u001b[39;49mh, units\u001b[39m=\u001b[39;49munits, dpi\u001b[39m=\u001b[39;49mdpi)\n\u001b[0;32m    112\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[0;32m    113\u001b[0m     x_scale, y_scale \u001b[39m=\u001b[39m save(scene, imgName, w\u001b[39m=\u001b[39mw, h\u001b[39m=\u001b[39mh, units\u001b[39m=\u001b[39munits, dpi\u001b[39m=\u001b[39mdpi)\n","File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python310\\site-packages\\ete3\\treeview\\main.py:752\u001b[0m, in \u001b[0;36msave\u001b[1;34m(scene, imgName, w, h, dpi, take_region, units)\u001b[0m\n\u001b[0;32m    750\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[0;32m    751\u001b[0m     targetRect \u001b[39m=\u001b[39m QRectF(\u001b[39m0\u001b[39m, \u001b[39m0\u001b[39m, w, h)\n\u001b[1;32m--> 752\u001b[0m     ii\u001b[39m=\u001b[39m QImage(w, h, QImage\u001b[39m.\u001b[39;49mFormat_ARGB32)\n\u001b[0;32m    753\u001b[0m     ii\u001b[39m.\u001b[39mfill(QColor(Qt\u001b[39m.\u001b[39mwhite)\u001b[39m.\u001b[39mrgb())\n\u001b[0;32m    754\u001b[0m     ii\u001b[39m.\u001b[39msetDotsPerMeterX(dpi \u001b[39m/\u001b[39m \u001b[39m0.0254\u001b[39m) \u001b[39m# Convert inches to meters\u001b[39;00m\n","\u001b[1;31mTypeError\u001b[0m: arguments did not match any overloaded call:\n  QImage(): too many arguments\n  QImage(size: QSize, format: QImage.Format): argument 1 has unexpected type 'float'\n  QImage(width: int, height: int, format: QImage.Format): argument 1 has unexpected type 'float'\n  QImage(data: bytes, width: int, height: int, format: QImage.Format): argument 1 has unexpected type 'float'\n  QImage(data: PyQt5.sip.voidptr, width: int, height: int, format: QImage.Format): argument 1 has unexpected type 'float'\n  QImage(data: bytes, width: int, height: int, bytesPerLine: int, format: QImage.Format): argument 1 has unexpected type 'float'\n  QImage(data: PyQt5.sip.voidptr, width: int, height: int, bytesPerLine: int, format: QImage.Format): argument 1 has unexpected type 'float'\n  QImage(xpm: List[str]): argument 1 has unexpected type 'float'\n  QImage(fileName: str, format: typing.Optional[str] = None): argument 1 has unexpected type 'float'\n  QImage(a0: QImage): argument 1 has unexpected type 'float'\n  QImage(variant: Any): too many arguments"]}],"source":["from ete3 import Tree, TreeStyle\n","\n","newick_tree = '((A:2,(B:2,C:3):5):5,D:4);'\n","t = Tree(newick_tree)\n","ts = TreeStyle()\n","ts.show_branch_length = True\n","t.render('%%inline', w=183, units='mm', tree_style=ts)"]},{"cell_type":"markdown","metadata":{},"source":["**Exercice 5 :** Reécrivez l'arbre suivant au format Newick puis afficher-le. Les nombres correspondent aux longueurs des branches :\n","![](input/tree.png)"]},{"cell_type":"code","execution_count":null,"metadata":{"trusted":true},"outputs":[{"ename":"TypeError","evalue":"arguments did not match any overloaded call:\n  QImage(): too many arguments\n  QImage(size: QSize, format: QImage.Format): argument 1 has unexpected type 'float'\n  QImage(width: int, height: int, format: QImage.Format): argument 1 has unexpected type 'float'\n  QImage(data: bytes, width: int, height: int, format: QImage.Format): argument 1 has unexpected type 'float'\n  QImage(data: PyQt5.sip.voidptr, width: int, height: int, format: QImage.Format): argument 1 has unexpected type 'float'\n  QImage(data: bytes, width: int, height: int, bytesPerLine: int, format: QImage.Format): argument 1 has unexpected type 'float'\n  QImage(data: PyQt5.sip.voidptr, width: int, height: int, bytesPerLine: int, format: QImage.Format): argument 1 has unexpected type 'float'\n  QImage(xpm: List[str]): argument 1 has unexpected type 'float'\n  QImage(fileName: str, format: typing.Optional[str] = None): argument 1 has unexpected type 'float'\n  QImage(a0: QImage): argument 1 has unexpected type 'float'\n  QImage(variant: Any): too many arguments","output_type":"error","traceback":["\u001b[1;31m---------------------------------------------------------------------------\u001b[0m","\u001b[1;31mTypeError\u001b[0m                                 Traceback (most recent call last)","Cell \u001b[1;32mIn[9], line 7\u001b[0m\n\u001b[0;32m      5\u001b[0m ts \u001b[39m=\u001b[39m TreeStyle()\n\u001b[0;32m      6\u001b[0m ts\u001b[39m.\u001b[39mshow_branch_length \u001b[39m=\u001b[39m \u001b[39mTrue\u001b[39;00m\n\u001b[1;32m----> 7\u001b[0m t\u001b[39m.\u001b[39;49mrender(\u001b[39m'\u001b[39;49m\u001b[39m%%\u001b[39;49;00m\u001b[39minline\u001b[39;49m\u001b[39m'\u001b[39;49m, w\u001b[39m=\u001b[39;49m\u001b[39m183\u001b[39;49m, units\u001b[39m=\u001b[39;49m\u001b[39m'\u001b[39;49m\u001b[39mmm\u001b[39;49m\u001b[39m'\u001b[39;49m, tree_style\u001b[39m=\u001b[39;49mts)\n","File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python310\\site-packages\\ete3\\coretype\\tree.py:1392\u001b[0m, in \u001b[0;36mTreeNode.render\u001b[1;34m(self, file_name, layout, w, h, tree_style, units, dpi)\u001b[0m\n\u001b[0;32m   1388\u001b[0m     \u001b[39mreturn\u001b[39;00m drawer\u001b[39m.\u001b[39mget_img(\u001b[39mself\u001b[39m, w\u001b[39m=\u001b[39mw, h\u001b[39m=\u001b[39mh,\n\u001b[0;32m   1389\u001b[0m                           layout\u001b[39m=\u001b[39mlayout, tree_style\u001b[39m=\u001b[39mtree_style,\n\u001b[0;32m   1390\u001b[0m                           units\u001b[39m=\u001b[39munits, dpi\u001b[39m=\u001b[39mdpi, return_format\u001b[39m=\u001b[39mfile_name)\n\u001b[0;32m   1391\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[1;32m-> 1392\u001b[0m     \u001b[39mreturn\u001b[39;00m drawer\u001b[39m.\u001b[39;49mrender_tree(\u001b[39mself\u001b[39;49m, file_name, w\u001b[39m=\u001b[39;49mw, h\u001b[39m=\u001b[39;49mh,\n\u001b[0;32m   1393\u001b[0m                             layout\u001b[39m=\u001b[39;49mlayout, tree_style\u001b[39m=\u001b[39;49mtree_style,\n\u001b[0;32m   1394\u001b[0m                               units\u001b[39m=\u001b[39;49munits, dpi\u001b[39m=\u001b[39;49mdpi)\n","File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python310\\site-packages\\ete3\\treeview\\drawer.py:111\u001b[0m, in \u001b[0;36mrender_tree\u001b[1;34m(t, imgName, w, h, layout, tree_style, header, units, dpi)\u001b[0m\n\u001b[0;32m    109\u001b[0m scene\u001b[39m.\u001b[39maddItem(scene\u001b[39m.\u001b[39mmaster_item)\n\u001b[0;32m    110\u001b[0m \u001b[39mif\u001b[39;00m imgName\u001b[39m.\u001b[39mstartswith(\u001b[39m\"\u001b[39m\u001b[39m%%\u001b[39;00m\u001b[39minline\u001b[39m\u001b[39m\"\u001b[39m):\n\u001b[1;32m--> 111\u001b[0m     imgmap \u001b[39m=\u001b[39m save(scene, imgName, w\u001b[39m=\u001b[39;49mw, h\u001b[39m=\u001b[39;49mh, units\u001b[39m=\u001b[39;49munits, dpi\u001b[39m=\u001b[39;49mdpi)\n\u001b[0;32m    112\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[0;32m    113\u001b[0m     x_scale, y_scale \u001b[39m=\u001b[39m save(scene, imgName, w\u001b[39m=\u001b[39mw, h\u001b[39m=\u001b[39mh, units\u001b[39m=\u001b[39munits, dpi\u001b[39m=\u001b[39mdpi)\n","File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python310\\site-packages\\ete3\\treeview\\main.py:752\u001b[0m, in \u001b[0;36msave\u001b[1;34m(scene, imgName, w, h, dpi, take_region, units)\u001b[0m\n\u001b[0;32m    750\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[0;32m    751\u001b[0m     targetRect \u001b[39m=\u001b[39m QRectF(\u001b[39m0\u001b[39m, \u001b[39m0\u001b[39m, w, h)\n\u001b[1;32m--> 752\u001b[0m     ii\u001b[39m=\u001b[39m QImage(w, h, QImage\u001b[39m.\u001b[39;49mFormat_ARGB32)\n\u001b[0;32m    753\u001b[0m     ii\u001b[39m.\u001b[39mfill(QColor(Qt\u001b[39m.\u001b[39mwhite)\u001b[39m.\u001b[39mrgb())\n\u001b[0;32m    754\u001b[0m     ii\u001b[39m.\u001b[39msetDotsPerMeterX(dpi \u001b[39m/\u001b[39m \u001b[39m0.0254\u001b[39m) \u001b[39m# Convert inches to meters\u001b[39;00m\n","\u001b[1;31mTypeError\u001b[0m: arguments did not match any overloaded call:\n  QImage(): too many arguments\n  QImage(size: QSize, format: QImage.Format): argument 1 has unexpected type 'float'\n  QImage(width: int, height: int, format: QImage.Format): argument 1 has unexpected type 'float'\n  QImage(data: bytes, width: int, height: int, format: QImage.Format): argument 1 has unexpected type 'float'\n  QImage(data: PyQt5.sip.voidptr, width: int, height: int, format: QImage.Format): argument 1 has unexpected type 'float'\n  QImage(data: bytes, width: int, height: int, bytesPerLine: int, format: QImage.Format): argument 1 has unexpected type 'float'\n  QImage(data: PyQt5.sip.voidptr, width: int, height: int, bytesPerLine: int, format: QImage.Format): argument 1 has unexpected type 'float'\n  QImage(xpm: List[str]): argument 1 has unexpected type 'float'\n  QImage(fileName: str, format: typing.Optional[str] = None): argument 1 has unexpected type 'float'\n  QImage(a0: QImage): argument 1 has unexpected type 'float'\n  QImage(variant: Any): too many arguments"]}],"source":["from ete3 import Tree, TreeStyle\n","\n","newick_tree = '((A:10,B:8,C:7):5,(D:9,(E:5,F:5):2):2):0;'\n","t = Tree(newick_tree)\n","ts = TreeStyle()\n","ts.show_branch_length = True\n","t.render('%%inline', w=183, units='mm', tree_style=ts)"]},{"cell_type":"markdown","metadata":{},"source":["**Exercice 6 :** Expliquer la relation de parenté entre $A$, $B$ et $C$ ? Qu'elles sont les hypothèses qui pourraient expliquer ce type d'embranchement dans un arbre ? Donner une réponse détaillée."]},{"cell_type":"markdown","metadata":{},"source":["Réponse : "]},{"cell_type":"markdown","metadata":{},"source":["### UPGMA\n","\n","L'algorithme UPGMA se base sur la matrice de distance entre les séquences. À chaque itération, les séquences avec la distance la plus faible sont regroupées puis une nouvelle matrice de distance est calculée avec le nouveau groupe. Cette étape est répétée jusqu'à n'avoir plus qu'un seul groupe. Par exemple, avec la matrice de distance entre les séquences $A$, $B$, $C$ et $D$ suivante :\n","\n","|   &nbsp;   | A | B | C | D |\n","|   -   | - | - | - | - |\n","| **A** | &nbsp; | &nbsp; | &nbsp; | &nbsp; |\n","| **B** | 4 | &nbsp; | &nbsp; | &nbsp; |\n","| **C** | 8 | 8 | &nbsp; | &nbsp; |\n","| **D** | 2 | 4 | 8 | &nbsp; |\n","\n","Les séquences $A$ et $D$ sont les plus proches ($distance(A,D)=2$). On les regroupe et on met à jour la matrice :\n","\n","|   &nbsp;   | (A, D) | B | C |\n","|   -   | - | - | - |\n","| **(A, D)** | &nbsp; | &nbsp; | &nbsp; |\n","| **B** | 4 | &nbsp; | &nbsp; |\n","| **C** | 8 | 8 | &nbsp; | &nbsp; |\n","\n","On regroupe maintenant $(A,D)$ et $B$ ($distance((A,D),B) = 4$) :\n","\n","|   &nbsp;   | ((A, D), B) | C |\n","|   -   | - | - |\n","| **((A, D), B)** | &nbsp; | &nbsp; |\n","| **C** | 8 | &nbsp; |\n","\n","Important : les nouvelles distances sont calculées en moyennant les distances entre les membres du nouveau groupe et des groupes non modifiés pondéré par le nombre d'UTOs dans chaque groupe. Avec $i$ et $j$ les deux membres du groupe nouvellement formé et k les groupes restant : $d_{ij,k} = \\frac{n_id_{ik}}{n_i + n_j}+ \\frac{n_jd_{jk}}{n_i + n_j}$. Par exemple avec la distance entre $((A, D), B)$ et $C$:\n","\n","$distance(((A, D), B), C) = (distance((A, D), C)*2 + distance(B, C)) \\mathbin{/} 3 = (8*2 + 8) \\mathbin{/} 3 = 8 $.\n","\n","L'arbre final écrit dans le format Newick est : $((A, D), B), C);$ \n","\n","Et avec les distances : $((A:1, D:1):1, B:2):2, C:4);$ \n","\n","**Exercice 7 :** Implémenter une version d'UPGMA qui calcule l'arbre au format Newick **avec les distances** puis appliquer votre algorithme aux données. "]},{"cell_type":"code","execution_count":10,"metadata":{"trusted":false},"outputs":[{"name":"stdout","output_type":"stream","text":["[[0 4 8 2]\n"," [4 0 8 4]\n"," [8 8 0 8]\n"," [2 4 8 0]]\n"]},{"data":{"text/plain":["2"]},"execution_count":10,"metadata":{},"output_type":"execute_result"}],"source":["def UPGMA(Mat):\n","    Min = M[1][0]\n","    for i in range(len(Mat[0])):\n","        for j in range(len(Mat[0])):\n","            if (M[i][j] < Min) and M[i][j] != 0:\n","                Min = M[i][j]\n","    return Min\n","\n","M = np.array([[0,4,8,2],[4,0,8,4],[8,8,0,8],[2,4,8,0]])\n","print(M)\n","UPGMA(M)"]},{"cell_type":"markdown","metadata":{},"source":["**Exercice 8 :** Quelles sont les hypothèses faites par UPGMA ? Semblent-elles respectées dans le cas présent ?"]},{"attachments":{},"cell_type":"markdown","metadata":{},"source":["Réponse : UPGMA fait plusieurs hypothèses sur les données :\n","\n","- Toutes les séquences évoluent à un taux de changement similaire.\n","- La distance entre deux groupes est proportionnelle à la durée depuis leur divergence.\n","- Aucune perte de branche n'est autorisée.\n","\n","Dans le cas présent, les données ne semblent pas respecter la première hypothèse, car il est peu probable que toutes les séquences évoluent à un taux de changement similaire. De plus, l'absence de pertes de branches peut également être remise en question, car il est possible que certaines séquences aient perdu des régions importantes au fil du temps.\n","\n","Cependant, malgré ces hypothèses simplificatrices, UPGMA reste un algorithme utile pour obtenir une approximation de l'arbre phylogénétique, en particulier lorsque les données sont limitées. Il est important de noter que d'autres méthodes plus avancées existent pour prendre en compte la complexité des données, telles que les méthodes bayésiennes et maximum de vraisemblance.\n"]},{"attachments":{},"cell_type":"markdown","metadata":{},"source":["----\n","## Enracinement de l'arbre\n","\n","Après avoir utilisé UPGMA pour réaliser votre arbre, l'enracinement s'est normalement fait au poids moyen. \n","\n","**Exercice 9 :** Quelle autre méthode est-il possible d'utiliser pour enraciner un arbre ? Pouvez-vous l'utilisez ici ? Si oui, afficher le nouvel arbre."]},{"cell_type":"markdown","metadata":{},"source":["**Réponse :** "]},{"cell_type":"markdown","metadata":{},"source":["----\n","## Neighbor-joining\n","\n","Le neighbor-joining est un autre algorithme permettant de calculer un arbre phylogénique à partir d'une matrice de distance. Il a l'avantage de faire moins d'hypothèse qu'UPGMA sur les données (elles ne sont plus forcément ultramétrique) et il donne donc de meilleurs arbres dans presque tout les cas. Vous trouverez un example d'application de cet algorithme [ici](http://www.evolution-textbook.org/content/free/tables/Ch_27/T11_EVOW_Ch27.pdf).\n","\n","**Exercice 10 :** Implémenter l'algorithme du neighbor-joining, appliquer-le aux données puis enraciner l'arbre."]},{"cell_type":"code","execution_count":7,"metadata":{"trusted":false},"outputs":[],"source":["# fonction pour trouver l'élément minimum dans une matrice\n","def find_min(mat):\n","    min_val = np.inf  # initialisation à une valeur très grande\n","    min_i, min_j = -1, -1  # initialisation à une valeur invalide\n","    for i in range(mat.shape[0]):\n","        for j in range(i):\n","            if mat[i, j] < min_val:\n","                min_val = mat[i, j]\n","                min_i, min_j = i, j\n","    return min_val, min_i, min_j\n","\n","# fonction pour calculer les distances moyennes entre un noeud i et tous les autres noeuds\n","def calc_mean_dist(mat, i):\n","    n = mat.shape[0]\n","    return np.sum(mat[i]) / (n-2)\n","\n","# fonction pour recalculer la matrice de distance après la fusion de deux noeuds i et j\n","def update_matrix(mat, i, j):\n","    n = mat.shape[0]\n","    new_mat = np.zeros((n-1, n-1))\n","    # copie des distances de la matrice originale\n","    for k in range(n):\n","        if k != i and k != j:\n","            new_mat[min(k,i), max(k,i)-1] = mat[k, min(i,j)] + mat[k, max(i,j)] - mat[min(i,j), max(i,j)]\n","    # calcul des nouvelles distances entre le nouveau noeud et tous les autres noeuds\n","    for k in range(n-2):\n","        new_mat[-1, k] = (mat[i, k] - mat[j, k]) / 2\n","    return new_mat\n","\n","# fonction principale neighbor joining\n","def neighbor_joining(mat):\n","    n = mat.shape[0]\n","    dist_mat = mat.copy()  # copie de la matrice de distance originale\n","    nodes = list(range(n))  # liste des noeuds\n","    node_names = list(map(str, range(n)))  # noms des noeuds (pour affichage)\n","    while n > 2:\n","        # calcul des distances moyennes entre tous les noeuds\n","        mean_dist_mat = np.zeros((n, n))\n","        for i in range(n):\n","            for j in range(i):\n","                mean_dist_mat[i, j] = mean_dist_mat[j, i] = calc_mean_dist(dist_mat, i) + calc_mean_dist(dist_mat, j)\n","        # recherche de la paire de noeuds à fusionner\n","        _, min_i, min_j = find_min(mean_dist_mat)\n","        # mise à jour de la matrice de distance et de la liste des noeuds\n","        new_mat = update_matrix(dist_mat, min_i, min_j)\n","        dist_mat = new_mat.copy()\n","        nodes.pop(max(min_i, min_j))\n","        nodes.pop(min(min_i, min_j))\n","        nodes.append(n-2)\n","        node_names.append(\"(\" + node_names.pop(min_i) + \":\" + str(dist_mat[min_i, -1] - calc_mean_dist(dist_mat, min_i)) + \",\" +\n","                          node_names.pop(min_j) + \":\" + str(dist_mat[min_j, -1] - calc_mean_dist(dist_mat, min_j)) + \")\")\n","        n -= 1\n","    # création de l'arbre final\n","    newick_str = \"(\" + node_names[0] + \":\" + str(dist_mat[0, 1]) + \",\" + node_names[1] + \":\" + str(dist_mat[0, 1]) + \");\"\n","    return newick_str\n","\n","    "]},{"cell_type":"markdown","metadata":{},"source":["----\n","## Bootstrap\n","\n","Le bootstrap est une méthode permettant de calculer la robustesse des branches d'un arbre. Il sagit de recréer un jeu de données artificiel en tirant n positions dans l'alignement des séquences avec remise, n étant la longueur de l'alignement. On recréer un arbre à partir de ces nouvelles données puis on le compare avec l'arbre obtenu avec les données réelles. Si une branche de l'arbre réelle est présente dans l'arbre artificiel, son bootstrap augmente de 1. On répètera la procédure 100 fois afin d'avoir un score sur 100 pour chaque branche.\n","\n","**Exercice 11 :** Calculer le bootstrap des branches des arbres obtenus avec UPGMA et neighbor-joining."]},{"cell_type":"code","execution_count":null,"metadata":{"trusted":false},"outputs":[],"source":["# À remplir"]},{"cell_type":"markdown","metadata":{},"source":["----\n","## Conclusion\n","\n","**Exercice 12 :** Quelles sont vos conclusions par rapport à l'arbre phylogénique de _smilodon_, _homotherium_ et _M. trumani_ ? Comparer les deux méthodes. Comment expliquer les caractéristiques morphologiques similaires entre ces trois espèces ? Une réponse détaillée est attendue."]},{"cell_type":"markdown","metadata":{"collapsed":true},"source":["Réponse :"]}],"metadata":{"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.10.10"},"toc":{"base_numbering":1,"nav_menu":{},"number_sections":true,"sideBar":true,"skip_h1_title":false,"title_cell":"Table of Contents","title_sidebar":"Contents","toc_cell":false,"toc_position":{},"toc_section_display":true,"toc_window_display":false},"varInspector":{"cols":{"lenName":16,"lenType":16,"lenVar":40},"kernels_config":{"python":{"delete_cmd_postfix":"","delete_cmd_prefix":"del ","library":"var_list.py","varRefreshCmd":"print(var_dic_list())"},"r":{"delete_cmd_postfix":") ","delete_cmd_prefix":"rm(","library":"var_list.r","varRefreshCmd":"cat(var_dic_list()) "}},"types_to_exclude":["module","function","builtin_function_or_method","instance","_Feature"],"window_display":false},"vscode":{"interpreter":{"hash":"ff839bb5509ec2f028e05148c597faef29288e097dd39519be0e28e7a237be07"}}},"nbformat":4,"nbformat_minor":2}
